www/BaseClass.js

```diff
@@ -1,181 +1,130 @@
+    return this[VARS_FIELD].hasOwnProperty(key) ? this[VARS_FIELD][key] : null;
+  },
+
+  set: function(key, value, noNotify) {
+    var prev = this.get(key);
+
+    this[VARS_FIELD][key] = value;
 
-  Object.defineProperty(self, "hashCode", {
-      value: hashCode,
-      writable: false
-  });
-  self.empty = function() {
-    for (var key in Object.keys(_vars)) {
-      _vars[key] = null;
-      delete _vars[key];
-    }
-  };
-
-  self.get = function(key) {
-    return key in _vars ? _vars[key] : null;
-  };
-  self.set = function(key, value, noNotify) {
-    var prev = _vars[key];
-    _vars[key] = value;
     if (!noNotify && prev !== value) {
-      self.trigger(key + "_changed", prev, value);
+      this.trigger(key + '_changed', prev, value);
     }
-  };
-  self.bindTo = function(key, target, targetKey, noNotify) {
-    targetKey = targetKey === undefined || targetKey === null ? key : targetKey;
-    self.on(key + "_changed", function(prevValue, newValue) {
-      target.set(targetKey, newValue, noNotify === true);
+
+    return this;
+  },
+
+  bindTo: function(key, target, targetKey, noNotify) {
+    var targetKey = targetKey || key;
+
+    this.on(key + '_changed', function(oldValue, value) {
+      target.set(targetKey, value, noNotify);
     });
-  };
+  },
 
-  self.trigger = function(eventName) {
-    if (!eventName || !_globalListeners[eventName] || !_globalListeners[eventName][hashCode]) {
-      return;
+  trigger: function(eventName) {
+    if (!eventName || !this[SUBSCRIPTIONS_FIELD][eventName]) {
+      return this;
     }
 
-    var args = [];
-    for (var i = 1; i < arguments.length; i++) {
-      args.push(arguments[i]);
-    }
-    var event = document.createEvent('Event');
-    event.initEvent(eventName, false, false);
-    event.mydata = args;
-    event.myself = self;
-
-    var callbacks = _globalListeners[eventName][hashCode];
-    for (var i = 0; i < callbacks.length; i++) {
-      callbacks[i].listener(event);
-    }
-  };
-  self.on = function(eventName, callback) {
-    if (!eventName || !callback || typeof callback !== "function") {
-      return;
+    var listeners = this[SUBSCRIPTIONS_FIELD][eventName];
+    var i = listeners.length;
+    var args = Array.prototype.slice.call(arguments, 1);
+
+    while (i--) {
+      listeners[i].apply(this, args);
     }
 
-    _globalListeners[eventName] = _globalListeners[eventName] || {};
+    return this;
+  },
 
-    var listener = function(e) {
-      if (!e.myself || e.myself !== self) {
-        return;
-      }
-      var evt = Object.create(e);
-      var mydata = evt.mydata;
-      delete evt.mydata;
-      delete evt.myself;
-      self.event = evt;
-      callback.apply(self, mydata);
+  on: function(eventName, listener) {
+    this[SUBSCRIPTIONS_FIELD][eventName] = this[SUBSCRIPTIONS_FIELD][eventName] || [];
+    var topic = this[SUBSCRIPTIONS_FIELD][eventName];
+    var index = topic.push(listener);
+    var self = this;
+
+    return function() {
```

Comment:
Could you tell me why do you return this function? 
```
function() {
  self.off(eventName, listener)
}
```

I think this is not common way...
```js
map.on(plugin.google.maps.event.MAP_READY, onMapReady)();
```